core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -425,15 +450,30 @@ protected void responseWithFailure(Throwable t) {
         protected void doRun() throws Exception {
             setPhase(task, "replica");
             assert request.shardId() != null : "request shardId must be set";
+            ReplicaResult result;
             try (Releasable ignored = acquireReplicaOperationLock(request.shardId(), request.primaryTerm())) {
-                shardOperationOnReplica(request);
-                if (logger.isTraceEnabled()) {
-                    logger.trace("action [{}] completed on shard [{}] for request [{}]", transportReplicaAction, request.shardId(),
-                        request);
-                }
+                result = shardOperationOnReplica(request);
             }
-            setPhase(task, "finished");
-            channel.sendResponse(TransportResponse.Empty.INSTANCE);
+            result.respond(new ActionListener<TransportResponse.Empty>() {
```

Comment:
can we make that listener an inner class with a name? it will be easier to read...